app/models/discussion_reader.rb

```diff
@@ -53,20 +68,41 @@ def discussion_reader_volume
   end
 
   private
+  def serialize_read_ranges(ranges)
+    # could merge adjancent ranges in the future
+    self.read_sequence_id_ranges = ranges.sort_by{|r| r.first}.map{|r| [r.first,r.last].join(',')}.join(' ')
+  end
 
-  def read_attributes(read_at)
-    @read_attributes ||= begin
-      read_items         = discussion.items.where('events.created_at <= ?', read_at)
-      read_salient_items = discussion.salient_items.where('events.created_at <= ?', read_at)
-      {
-        last_read_at:             read_at,
-        last_read_sequence_id:    read_items.maximum(:sequence_id).to_i,
-        read_items_count:         read_items.count,
-        read_salient_items_count: read_salient_items.count
-      }
+  def parse_read_ranges
+    self.read_sequence_id_ranges.to_s.split(' ').map do |pair|
+      Range.new *pair.split(',').map(&:to_i)
     end
   end
 
+  def compute_read_items_count
+    parse_read_ranges.map(&:count).sum
+  end
+
+  def can_merge?(a, b)
+    a.include?(b.first - 1) || a.include?(b.last + 1)
+  end
+
+  def merge(ranges)
+    sorted = ranges.map{|r| [r.first, r.last] }.flatten.sort
```

Comment:
I might be tempted to make this
```
def merge(*ranges)
  ranges.flatten.map { |r| [r.first, r.last] }.sort
end
```

which makes `merge` an easier method to call
```
merge(mergeable, a, b, c)
```

Reply:
Also, I feels to me a bit like there may be an object trying to get out here
```
class DiscussionReader
  before_save :consolidate_ranges
  def ranges
    @ranges ||= RangeCollection.new(self.read_sequence_id_ranges)
  end

  def consolidate_ranges
    ranges.reduce.to_s
  end
end

class RangeCollection
  def initialize(rang_str)
    # maybe this takes the string from the database directly and converts it to an array of Ranges internally?
    @ranges = range_str.split(' ').map { |r| Range.new(r.split(',')[0], r.split(',')[1]) }
  end

  def reduce # (from reduce_ranges)
    # returns a new SequenceRangeCollection which has compacted @ranges so that adjacent ranges are merged
  end

  def sorted
    # sort @ranges so they're in order
  end

  def to_s
    # takes the array of ranges and converts them to database-friendly format
    @ranges.map { |r| [r.first, r.last].join(',') }.join(' ')
  end
end
```